src/net/addr.rs

```diff
@@ -0,0 +1,302 @@
+use crate::future::Future;
+use crate::task::blocking::JoinHandle;
+use futures::future::{ready, Ready};
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+pub use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};
+
+/// A trait for objects which can be converted or resolved to one or more
+/// [`SocketAddr`] values.
+///
+/// This trait is an async version of [`std::net::ToSocketAddrs`].
+///
+/// [`std::net::ToSocketAddrs`]: https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html
+pub trait ToSocketAddrs {
+    /// Returned iterator over socket addresses which this type may correspond
+    /// to.
+    type Iter: Iterator<Item = SocketAddr> + Send;
+    /// Returned Future
+    type Output: Future<Output = crate::io::Result<Self::Iter>> + Send;
+    /// Converts this object to an iterator of resolved `SocketAddr`s.
+    ///
+    /// The returned iterator may not actually yield any values depending on the
+    /// outcome of any resolution performed.
+    ///
+    /// Note that this function may block a backend thread while resolution is
+    /// performed.
+    fn to_socket_addrs(&self) -> Self::Output;
```

Comment:
This method would ideally be an `async fn`, which is currently a feature supported by the language. So it'd be best if we could design the API in such a way that switching to `async fn` is not a breaking change once that becomes possible.

I suggest replacing the `Output` associated type with a hack similar to `async_std::stream::Stream` and other traits in the library, where we return a hidden concrete type, but "lie" it's actually an `impl Future`  in the generated docs.

Reply:
Pushed some code for this, there is a different with other `ret!`, but I think that's ok.